Add Error-Handling, Explaination (Needs a grammar check), Improved Code & Added gitingore file

VeraDev0 opened the pull request:
Ive improved the DLL mod a lot further by adding the following elements:
- **Error Handling:** Tells you if something gets screwed.
- **Game Version Check:** Tells you if the DLL would not work in the game version your currently in.
- **Explanation:** Ive gone and looked how it worked and provided a explanation on how the DLL mod works and how to make it work for future game updates.
- **Changed how it patches the game:** Now uses memcpy instead of writing to a address directly.

The only problem right now is my poor English meaning i suggest to modify the Explanation comments to have proper English, other than that, basically a full revamp.

VeraDev0 committed:
Add Error-handling, explaining how it works and using memcpy.

Added error-handling to tell you if it failed.
Added explaination over how the DLL works and how to manage it for future game updates.
Use memcpy instead of directly setting it.

VeraDev0 committed:
Add Git ignore file & Fix newAddress

VeraDev0 committed:
Delete NetworkChecksumDisabler.vcxproj.user

VeraDev0 committed:
Fix git ingore

VeraDev0 committed:
Forgot to uncomment some Error-handling

crackx02 commented:
I see a couple slight improvements you could do before this is merged:
- Address Vectors: Using an `std::vector` makes a heap allocation, which is unnecessary for the 4 bytes of data in this case. Consider using a `constexpr std::array` instead, which avoids this.
- Similar to the vectors, the `address` integer could also be made constexpr instead of static. This makes sure it is evaluated at compile time.
- `const uint64_t& base = (uint64_t)GetModuleHandle(NULL);`: This assigns a const reference to a temporary integer. It probably works but is kind of weird code and the reference isn't needed here. Please remove the `&` from the variable.

VeraDev0 commented:
I never knew about std::array so thats why ive used a vector.

Didnt know anything about constexpr, likey have to research on it.

The reference is there because in my coding style, i const and reference everything (for optimization reasons) but some arent because they might not be referencable or needs to be modified.

BenMcAvoy commented:
`std::array` is similar to a standard array but it is further typesafe and has useful functions such as `size()`, it is similar to the `std::vector` API but it is not on the heap (instead stack allocated) and is completely fixed in size and supports bounds checking with `.at()`.

Also, when using `constexpr` it can be used in two scenarios:
- On a function
- On some data

On a function, it will attempt to run that function at compile time and template in the result to wherever the function is called throughout the code. You can imagine a simple fibonacci function being marked as `constexpr` and at compile time, it will actually call this function and template in the result of any calls to it. You are essentially telling the compiler the function should be evaluated at compile time. It is not a guarantee but is a hint to the compiler and the compiler will decide whether this "contract" between you and the compiler is valid (whether it follows rules that allow constexpr)

On some data, it does something similar but of course does not need to call a function as it already has the data.

I am by no means an expert and am still young so apologies for any mistakes, but hopefully this helps.

BenMcAvoy commented:
I will also post an updated explanation here of how the DLL works (hopefully today)

VeraDev0 commented:
Ive applied the changes like what you said, i also removed some code due to being not needed.
Pretty sure this meets everything now. (Unless the explanation has to be updated)

Its actually a lot more readable than when i made this pull request.

VeraDev0 committed:
Use std::array, remove unnessorary code and etc

Using std::array to not do a heap allocation
Address is now a constexpr
The base address is nolonger referenced
Removed unnessorary code (mostly error-handling)

Because of these changes, its actually quite more readable now.

BenMcAvoy commented:
Looks great, what do you mean about the description though, I don't see an update. Hopefully when I get home I can upload the one I modified to have correct grammar

BenMcAvoy commented:
Just as a warning, this is unimportant and simply a curiosity of mine.

Why does your coding style mandate that you define variables that are assigned to a function result that returns an actual value? I would normally only use it to reference to a value in an unordered map, something similar to that or any kind of binding where I essentially want an alias, using a reference for performance reasons. 

I'm mostly confused why you used it here due to the fact that either way the variable is 64 bits (uint64_t versus a 64 bit reference)

Purely interest here, am I missing some C++ magic?

VeraDev0 commented:
> Just as a warning, this is unimportant and simply a curiosity of mine.
> 
> Why does your coding style mandate that you define variables that are assigned to a function result that returns an actual value? I would normally only use it to reference to a value in an unordered map, something similar to that or any kind of binding where I essentially want an alias, using a reference for performance reasons.
> 
> I'm mostly confused why you used it here due to the fact that either way the variable is 64 bits (uint64_t versus a 64 bit reference)
> 
> Purely interest here, am I missing some C++ magic?

Its just C++ magic and its weirdness aspect. Here's a example

These do might do the same at the high-level but don't do the same in the assembly side.
```cpp
i++
++i
```
They achieve the same result however, `++i` is actually a lot faster. Don't ask me why but its like that.
And another thing is that putting `const` and `&` literally everywhere where possible + it is also a optimization* + I find the code a lot more readable like that. 

There are a lot more C++ magic i can tell you like that this is the fastest way to iterate over vector's and maps. (Generally very large ones)
```cpp
const std::vector<std::string>& myVec = {"String A", "Lorem ispum", "Yes", "Just pretend there is more data here"};

const auto& start = myVec.begin();
const auto& end = myVec.end();

for (auto it = start; it != end; ++it)
    std::cout << *it << "\n";
```

For proof, i have ran a benchmark which would loop through 10000 randomly generated numbers and print them all, `Optimized loop` is the one i've mentioned and `Unoptimized Loop` is just doing it via `for(uint8_t number : numberVec)`

I also did a variation with Optimized Loop without references and consts

Results:
```
[Timer_t]: "Optimized Loop" Elapsed Time: 315ms
[Timer_t]: "No Consts & References - Optimized Loop" Elapsed Time: 331ms
[Timer_t]: "Unoptimized Loop" Elapsed Time: 475ms
```

It might not look like a massive difference but i prefer super fast code and so this is why i do references and consts everywhere.

BenMcAvoy commented:
Good point, I still find it funny how much faster iterators are and how unintegrated their syntax is... does the compiler not optimise ++i and i++ to the same thing in Release or is this another hole in MSVC? (I come mainly from Clang/GCC where as far as I was concerned optimisation flags fixed this issue)

I still don't understand *in this case* at least why `const xxx&` would be faster as either way it is 64 bits. However yes I understand in other scenarios it would of course be much faster (e.g. pulling a reference to a struct stored as a value in an unordered map or some other type which is much larger than 64 bits)

VeraDev0 commented:
I am pretty sure it doesn't optimize it

`const xxx&` is faster for a lot of things (not just for numbers),  but I'm not sure why tho, but I'm guessing `const` would make it faster since it would be never modified. Not sure about the reference part but it does make it go faster so i just use it everywhere where possible.

If your wondering, i learned this from one of my friends who has more experience & knowledge in C++ than me and makes code go way too fast.

BenMcAvoy commented:
Yeah I know who you're talking about haha. Usually I've seen Const to support R-Values in your references and also of course makes your software more stable as a whole due to unintentional side effects being harder to cause

References definitely are faster when using a data type larger than 64 bits

BenMcAvoy commented:
I was simply confused why you would have a reference to a number especially when it's the same size      

I may check later whether it's even possible to do this without const 🤔 (due to it being an rvalue you were referencing essentially)
